### PR TITLE
perf(rtp): reduce rtp scans

### PR DESCRIPTION
**Pull request by lewis6991**
## Problem:
Scanning the filesystem is expensive and particularly affects startuptime.

## Solution:

Reduce the amount of redundant directory scans.

Change `DoInRuntimeCB` to be able to handle multiple matches so callbacks can sort and prioritize.

### CODE REVIEW

**Comment by lewis6991** [comment]
This essentially does *some* of the optimizations done in lazy.nvim.

**Comment by clason** [comment]
Out of interest: does this affect (help solve) https://github.com/neovim/neovim/pull/23202 and friends?

**Comment by lewis6991** [comment]
Not sure. It does change things slightly.

Before if you had:
```
a/
    bar.vim
    baz.lua
b/
    qux.vim
    quux.lua
```

would be sourced in this order (all vim, the all lua):

```
a/bar.vim
b/aux.vim
a/baz.lua
b/quux.lua
```

with this PR:

```
a/bar.vim
a/baz.lua
b/qux.vim
b/quux.lua
```

There is still priority given to vim files, but this is done per directory.

**Comment by clason** [comment]
> would be sourced in this order (all vim, the all lua):

I thought I fixed that... but on the higher level, so this change may indeed allow us to simplify the `runtime!` calls across the codebase.

File: src/nvim/runtime.c
Comment (bfredl) [inline src/nvim/runtime.c:854]: perhaps `expand_wildcards_and_source` and ? This usecase seems to be the implied one given that it hardcodes matching of .vim then .lua files

Comment (lewis6991) [reply src/nvim/runtime.c:854]: This is also called with `add_pack_start_dir` and `add_start_pack_plugin` which in those cases the first two loops (for `.lua` and `.vim`) don't do anything.

Comment (lewis6991) [reply src/nvim/runtime.c]: Modifying the signature of `DoInRuntimepathCB` to support multiple files seems to be a better route.

File: runtime/syntax/synload.vim
Comment (clason) [inline runtime/syntax/synload.vim:54]: Just to be clear: the ordering "Vim first, then Lua" is now fully taken care of by `runtime!`? Otherwise you need to interleave Vim and Lua files as in #24147 

(also check `ftplugin.vim` and `indent.vim`, cf. #20322)

Comment (clason) [reply runtime/syntax/synload.vim:54]: Oh, and do these changes also apply to `nvim__get_runtime`? Could try using this instead of `runtime!` here, then.

Comment (lewis6991) [reply runtime/syntax/synload.vim:54]: Yes I believe so. Any pattern passed to `runtime` will source vim files first.

Comment (clason) [reply runtime/syntax/synload.vim:54]: ...but making sure to respect directory ordering

Comment (lewis6991) [reply runtime/syntax/synload.vim:54]: Not, this is now the same style that `ftplugin.vim` does not **currently**. Not sure if that was intentional:

```vim
      " When there is a dot it is used to separate filetype names.  Thus for
      " "aaa.bbb" load "aaa" and then "bbb".
      for name in split(s, '\.')
        " Load Lua ftplugins after Vim ftplugins _per directory_
        " TODO(clason): use nvim__get_runtime when supports globs and modeline
        exe printf('runtime! ftplugin/%s.vim ftplugin/%s.lua', name, name)
        exe printf('runtime! ftplugin/%s_*.vim ftplugin/%s_*.lua', name, name)
        exe printf('runtime! ftplugin/%s/*.vim ftplugin/%s/*.lua', name, name)
      endfor
```

Need to double-check how `:runtime` handles each argument; does each argument do a seperate scan(?).

Comment (lewis6991) [reply runtime/syntax/synload.vim:54]: Ok, currently each argument to `:runtime!` performs a separate scan, so the change in this file shouldn't do anything.

```vim
:runtime! a
:runtime! b
```
should be the same as
```
runtime! a b
```

Comment (clason) [reply runtime/syntax/synload.vim:54]: > Note, this is now the same style that `ftplugin.vim` does **currently**. Not sure if that was intentional:
> 
> ```viml
>       " When there is a dot it is used to separate filetype names.  Thus for
>       " "aaa.bbb" load "aaa" and then "bbb".
>       for name in split(s, '\.')
>         " Load Lua ftplugins after Vim ftplugins _per directory_
>         " TODO(clason): use nvim__get_runtime when supports globs and modeline
>         exe printf('runtime! ftplugin/%s.vim ftplugin/%s.lua', name, name)
>         exe printf('runtime! ftplugin/%s_*.vim ftplugin/%s_*.lua', name, name)
>         exe printf('runtime! ftplugin/%s/*.vim ftplugin/%s/*.lua', name, name)
>       endfor
> ```
> 
> Need to double-check how `:runtime` handles each argument; does each argument do a seperate scan(?).

No, the current state of `ftplugin.vim` is **wrong** (I messed up); #24147 should be the right behavior (I hope; still waiting for confirmation).

File: src/nvim/runtime.c
Comment (zeertzjq) [inline src/nvim/runtime.c:290]: This may change behavior if someone uses `:runtime!` to load alphabetically-sorted files (e.g. `00_foo.lua`, `10_foo.vim`, `20_foo.lua`, `30_foo.vim`). Not sure if anyone is using this with multiple extensions though.

File: src/nvim/runtime.c
Comment (justinmk) [inline src/nvim/runtime.c:847]: could live in strings.c near `vim_strchr`

**Review by justinmk — APPROVED** [review]
Nice!

File: src/nvim/runtime.c
Comment (zeertzjq) [reply src/nvim/runtime.c:290]: Ah yes. This changes the behavior of `:runtime! plugin/**/*.\(vim\|lua\)` to be the same of `:runtime! plugin/**/*.vim plugin/**/*.lua`.

Comment (justinmk) [reply src/nvim/runtime.c:290]: I assumed that was intentional. Otherwise all of these `do_source` calls could be in a single for-loop.

Comment (zeertzjq) [reply src/nvim/runtime.c:290]: I think this behavior change needs to be documented. To achieve the old behavior one has to expand the pattern and the source the files one by one.

Comment (clason) [reply src/nvim/runtime.c:290]: In any case, we should decide and unambiguously document the order. (Which this PR is already improving.)

Comment (lewis6991) [reply src/nvim/runtime.c:290]: We don't want to lock down the implementation too much without good reason. I think all we need to be strict on is that any `foo/bar.vim` is loaded before `foo/bar.lua`. Anything outside of that should be explicitly undefined.

Comment (zeertzjq) [reply src/nvim/runtime.c:290]: > I think all we need to be strict on is that any `foo/bar.vim` is loaded before `foo/bar.lua`.

This should still be documented at `:h runtime`, as previously `:runtime` doesn't have such special-cased behavior.

Comment (clason) [reply src/nvim/runtime.c:290]: > We don't want to lock down the implementation too much without good reason. I think all we need to be strict on is that any `foo/bar.vim` is loaded before `foo/bar.lua`. Anything outside of that should be explicitly undefined.

Then we should document _that_. But I do think it's important to know whether `foo/bar.lua` is loaded before `foo/baz.vim` or not.

Comment (lewis6991) [reply src/nvim/runtime.c:290]: Added something to `:h runtime`.

File: runtime/doc/starting.txt
Comment (clason) [inline runtime/doc/starting.txt:514]: As written, this kinda contradicts the sentence in `repeat.txt` -- here it sounds like "all vim files, then all lua files"?

I'd recommend writing one (longer) sentence and just copy-pasting it in both places, to make sure they're consistent.

Comment (lewis6991) [reply runtime/doc/starting.txt:514]: Well currently this is what this PR implements, but yes, this needs to be a looser description to match what we have for `:runtime`.

Comment (clason) [reply runtime/doc/starting.txt:514]: To be clear, I'm fine with either behavior (until presented with a compelling concrete and actual example where one would be breaking); I'd just prefer _stricter_  and _matching_ description of the behavior implemented here.

File: runtime/doc/repeat.txt
Comment (zeertzjq) [inline runtime/doc/repeat.txt:236]:
```suggestion
			For each {file} pattern, if both `.vim` and `.lua`
			files match in a directory, the `.vim` files take
			priority and are sourced first.
```

Comment (lewis6991) [reply runtime/doc/repeat.txt:236]: This is stronger than what is intended. We just want to enforce that `foo/bar.vim` take priority over `foo/bar.lua`.

Comment (zeertzjq) [reply runtime/doc/repeat.txt:236]: At least add the "in a directory"

Comment (zeertzjq) [reply runtime/doc/repeat.txt:236]: How about this
```suggestion
			For each {file} pattern, if a multiple matches differ
			only in extension, the `.vim` version is sourced first
			the `.lua` version second, and others last.
```

**Comment by clason** [comment]
> with this PR:
> 
> ```
> a/bar.vim
> a/baz.lua
> b/qux.vim
> b/quux.lua
> ```

I think such an example would be useful in the docs (in one place only) as this would obviate a need for a long explanation, but should include a `a/quux.vim` (so you can distinguish "per filename" and "per directory").

**Comment by lewis6991** [comment]
This is not behaviour we need/want to guarantee. In this example, the order these are sourced in is officially undefined (even if it's implemented to do something defined).

**Comment by clason** [comment]
Well, I don't think this was _intended_ to be "undefined behavior" -- this just happens to be "undocumented behavior", which is of course convenient for us since we can shrug our shoulders whenever someone complains. I don't think that's a good stance though. I do believe that at the least we should remark than any other order is not guaranteed (even if observed) and should not be relied on.

(We had enough confused reports about this which were unclear whether this was an issue.)

**Comment by lewis6991** [comment]
This PR is only intended to improve performance. Changes to documented behaviour that lock down implementation can be done in follow up PR's.

Just to note, lazy.nvim, which is what inspired this PR, doesn't give any priority to `.vim` files, but does source them alphabetically which means it prioritises lua files ([ref](https://github.com/folke/lazy.nvim/blob/5af331ea65418dc9361769891921fdee4bcc837a/lua/lazy/core/loader.lua#L410-L426)). Not a big problem, and easy to fix, but just pointing out that guaranteeing this behaviour doesn't give us much value (IMO), especially given the decline of vimscript.

**Review by zeertzjq — APPROVED** [review]

**Comment by zeertzjq** [comment]
Looks like there are some problems with this approach: #24292

**Comment by lewis6991** [comment]
Sorry can you help me put one and one together? I don't fully follow.

File: src/nvim/runtime.c
Comment (justinmk) [inline src/nvim/runtime.c:288]: Is it not worth the code simplification to combine all these into the same for-loop, and would that preserve the "old" behavior?

**Comment by zeertzjq** [comment]
#24296

File: runtime/doc/starting.txt
Comment (zeertzjq) [inline runtime/doc/starting.txt:509]:
```suggestion
		:runtime! plugin/**/*.{vim,lua}
```

File: src/nvim/runtime.c
Comment (zeertzjq) [inline src/nvim/runtime.c:1034]:
```suggestion
  static const char *ftpat = "%s/ftdetect/*.{vim,lua}";  // NOLINT
```

File: src/nvim/runtime.c
Comment (zeertzjq) [inline src/nvim/runtime.c:1040]:
```suggestion
  vim_snprintf(pat, len, "%s/plugin/**/*.{vim,lua}", ffname);  // NOLINT
```

File: src/nvim/runtime.c
Comment (zeertzjq) [inline src/nvim/runtime.c:1171]:
```suggestion
    char *const plugin_pattern = "plugin/**/*.{vim,lua}";  // NOLINT
```

**Review by zeertzjq — COMMENTED** [review]
After #24296

File: src/nvim/runtime.c
Comment (lewis6991) [reply src/nvim/runtime.c:1171]: I've updated the change so the code doesn't rely on suffix patterns and the vim and lua case now has a specialized path.

**Comment by lewis6991** [comment]
This is probably worth a re-review.

I've now also reduced scans in `color/*` and `compiler/*`.

This lead to a behaviour change (bugfix?), where `compiler/*.lua` will always execute, whereas before the presence of vim files would prevent this. However, this didn't make sense since `DIP_ALL` was passed which indicates all found files should be sourced.

**Comment by zeertzjq** [comment]
Yes I notice the bug with compiler before.

File: src/nvim/highlight_group.c
Comment (zeertzjq) [inline src/nvim/highlight_group.c:704]: Combine these two calls into one call with `DIP_START + DIP_OPT`

File: src/nvim/runtime.c
Comment (zeertzjq) [inline src/nvim/runtime.c:1092]: ditto

File: src/nvim/highlight_group.c
Comment (lewis6991) [reply src/nvim/highlight_group.c:704]: What about `DIP_NORTP`? I don't fully understand what that flag does.

Comment (zeertzjq) [reply src/nvim/highlight_group.c]: Not needed anymore.

File: runtime/doc/repeat.txt
Comment (justinmk) [inline runtime/doc/repeat.txt:236]:
```suggestion
			and `.lua` extensions, the `.vim` version will be sourced first.
```

"Priority" is whatever is sourced last, right? So maybe it's less confusing to not mention priority here.

File: src/nvim/runtime.c
Comment (justinmk) [inline src/nvim/runtime.c:885]: is `source_in_path` needed anymore? Could rename this if not. 
```suggestion
/// Like source_runtime(), but:
///   - use "path" instead of 'runtimepath'
///   - only source .vim and .lua files
int source_in_path(char *path, char *name, int flags)
```

File: src/nvim/runtime.c
Comment (justinmk) [inline src/nvim/runtime.c:272]:
```suggestion
/// Source all .vim and .lua files in "fnames". If both "x.vim" and "x.lua" are found, the vim file is sourced first.
```

Comment (lewis6991) [reply src/nvim/runtime.c:272]: This isn't quite true since if `x.vim`, `y.vim`, `x.lua` and `y.lua` are found they are sourced in that order.

File: src/nvim/runtime.c
Comment (lewis6991) [reply src/nvim/runtime.c:885]: Nope 😄

**Review by justinmk — APPROVED** [review]

File: src/nvim/runtime.c
Comment (zeertzjq) [inline src/nvim/runtime.c:273]:
```suggestion
/// Source all .vim and .lua files in "fnames" with .vim files being sourced first.
```

File: src/nvim/runtime.c
Comment (zeertzjq) [inline src/nvim/runtime.c:301]:
```suggestion
/// Source all files in "fnames" with .vim files being sourced first.
```

**Review by zeertzjq — APPROVED** [review]

**Pull request merged**
